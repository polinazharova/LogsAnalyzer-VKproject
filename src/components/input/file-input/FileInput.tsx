import { ChangeEventHandler } from "react";

interface FileInputProps {
  children: React.ReactNode;
  formClass: string;
  inputId: string;
  handleFileChange: ChangeEventHandler<HTMLInputElement>;
}

const FileInput: React.FC<FileInputProps> = ({
  children,
  formClass,
  inputId,
  handleFileChange,
}) => {
  return (
    <div className={`${formClass}-file-input-comp`}>
      <label className={`${formClass}-file-input-label`} htmlFor={inputId}>
        {children}
      </label>
      <input
        className={`${formClass}-file-input`}
        id={inputId}
        name={inputId}
        type="file"
        multiple
        accept=".log"
        onChange={handleFileChange}
      ></input>
    </div>
  );
};

export default FileInput;
