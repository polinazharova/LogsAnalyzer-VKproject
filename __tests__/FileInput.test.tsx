import React from "react";
import FileInput from "../src/components/input/file-input/FileInput";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

describe("File input component", () => {
  test("renders children correctly", () => {
    render(
      <FileInput formClass="test" inputId="test" handleFileChange={() => {}}>
        Input
      </FileInput>,
    );
    expect(screen.getByText("Input")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <FileInput formClass="test" inputId="test" handleFileChange={() => {}}>
        Input
      </FileInput>,
    );
    expect(container).not.toBeEmptyDOMElement();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("applies correct className, inputId", () => {
    render(
      <FileInput formClass="test" inputId="test" handleFileChange={() => {}}>
        Input
      </FileInput>,
    );
    const fInput = screen.getByLabelText("Input");

    expect(fInput).toHaveClass("test-file-input");
    expect(fInput.id).toBe("test");
  });

  test("uploads a .log file", async () => {
    const user = userEvent.setup();
    const file = new File(
      [
        "2023-01-01 [ERROR] Something went wrong\n2023-01-02 [INFO] Process started",
      ],
      "demo.log",
      { type: "text/plain" },
    );

    render(
      <FileInput formClass="test" inputId="test" handleFileChange={() => {}}>
        Input
      </FileInput>,
    );
    const fInput = screen.getByLabelText("Input") as HTMLInputElement;

    await user.upload(fInput, file);

    expect(fInput.files).toHaveLength(1);
    expect(fInput.files?.[0]).toBe(file);
  });

  test("calls onChange handler when a file chosed", async () => {
    const handleFileChange = jest.fn();
    const user = userEvent.setup();
    const file = new File(
      [
        "2023-01-01 [ERROR] Something went wrong\n2023-01-02 [INFO] Process started",
      ],
      "demo.log",
      { type: "text/plain" },
    );

    render(
      <FileInput
        formClass="test"
        inputId="test"
        handleFileChange={handleFileChange}
      >
        Input
      </FileInput>,
    );
    const fInput = screen.getByLabelText("Input") as HTMLInputElement;

    await user.upload(fInput, file);

    expect(handleFileChange).toHaveBeenCalledTimes(1);
  });
});
