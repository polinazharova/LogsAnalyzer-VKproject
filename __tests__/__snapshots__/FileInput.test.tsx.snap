// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File input component matches snapshot 1`] = `
<div
  class="test-file-input-comp"
>
  <label
    class="test-file-input-label"
    for="test"
  >
    Input
  </label>
  <input
    accept=".log"
    class="test-file-input"
    id="test"
    multiple=""
    name="test"
    type="file"
  />
</div>
`;
